portfolio/src/test/java/com/google/sps/servlets/CommentsServletTest.java

```diff
@@ -26,16 +43,127 @@ public void setUp() {
+    request.addParameter("text", "A nice comment");
+    try {
+      servlet.doPost(request, response);
+      assertEquals(response.getStatus(), REDIRECT);
+      String responseString = response.getContentAsString();
+      JsonObject jsonObject = (new JsonParser()).parse(responseString).getAsJsonObject();
+      assertEquals(jsonObject.get("username").toString(), quoted("Paul"));
+      assertEquals(jsonObject.get("text").toString(), quoted("A nice comment"));
+    } catch (IOException ex) {
+      System.out.println(ex.getMessage());
+    }
+  }
+
+  @Test
+  public void testCreateWithValidPayloadAndValidReplyKey() {
+    // create a comment and put it to the datastore
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+    Entity commentEntity = new Entity("Comment");
+    ds.put(commentEntity);
+    commentEntity.setProperty("username", "paul");
+    commentEntity.setProperty("text", "Very cool comment text");
+    request.setMethod("POST");
+    request.setContentType("text/html");
+    request.addParameter("username", "Paul");
+    request.addParameter("text", "A nice comment");
+    // add a valid reply to field
+    String replyToKey = KeyFactory.keyToString(commentEntity.getKey());
+    request.addParameter("replyTo", replyToKey);
+    try {
+      servlet.doPost(request, response);
+      assertEquals(response.getStatus(), REDIRECT);
+      String responseString = response.getContentAsString();
+      JsonObject jsonObject = (new JsonParser()).parse(responseString).getAsJsonObject();
+      assertEquals(jsonObject.get("username").toString(), quoted("Paul"));
+      assertEquals(jsonObject.get("text").toString(), quoted("A nice comment"));
+      assertEquals(jsonObject.get("replyTo").toString(), quoted(replyToKey));
+    } catch (IOException ex) {
+      System.out.println(ex.getMessage());
+    }
+  }
+
+  @Test
+  public void testCreateWithNoPayload() {
+    request.setMethod("POST");
+    request.setContentType("text/html");
+    try {
+      servlet.doPost(request, response);
+      assertEquals(response.getStatus(), BAD_REQUEST);
+    } catch (IOException ex) {
+      System.out.println(ex.getMessage());
+    }
+  }
+
+  @Test
+  public void testCreateWithAbsentCommentText() {
+    request.setMethod("POST");
+    request.setContentType("text/html");
+    // set the username only, without the text
+    request.addParameter("username", "Paul");
+    try {
+      servlet.doPost(request, response);
+      assertEquals(response.getStatus(), BAD_REQUEST);
+    } catch (IOException ex) {
+      System.out.println(ex.getMessage());
+    }
+  }
+
+  @Test
+  public void testCreateWithInvalidCommentTextValue() {
     request.setMethod("POST");
     request.setContentType("text/html");
-    request.addParameter("username", "paul");
+    request.addParameter("username", "Paul");
+    // set the username and text full of spaces
+    request.addParameter("text", "        ");
```

Comment:
I am not aware of the other PRs so I don't know how you implemented this. But checking for white space on backend seems like unnecessary overhead. You can very easily check this on frontend itself and not let the POST happen in the first place.

Reply:
My frontend does not check for empty input because it uses plain HTML form

Reply:
You can definitely do it even with HTML5 and java script together. Check out https://developer.mozilla.org/en-US/docs/Learn/Forms/Form_validation#Validating_forms_using_JavaScript for example. Infact, you can even write your own custom validator in js.

Reply:
Thank you, didnt know about this

Reply:
So do you suggest to remove this test as well as check for empty input from servlet?